### ISSUE TITLE
Support for direct rendering to Framebuffer

### ISSUE THREAD

**Issue by shwetashah16** [ardera/flutter-pi#457]
Is there a way to override rendering to use direct frame buffer? I am using armv7 32 bit device which only has support for Framebuffers.

**Comment by ardera** [comment]
Currently, no, there is some basic infrastructure there so it could be implemented, but just hasn't been done yet. Do you mind telling me what device you're talking about precisly? iMX6?

**Comment by shwetashah16** [comment]
Device is in-house created based on Qualcomm chip set sdx65, which we want to try replace current display UI with Flutter.

**Comment by gotnull** [comment]
There is a way to get framebuffer output, just not from inside flutter-pi: go through the Flutter embedder API directly, use its software renderer, and copy the CPU buffer to `/dev/fb0` yourself.

Mine is an arm64 Allwinner H700 rather than your armv7 part, so not identical, but the device likewise has no usable DRM path and the approach leans on no GPU at all, so it should carry over to a framebuffer-only armv7 target. The render loop is just mmap the fb and blit each frame (the engine buffer is BGRA, so mind the channel order against your fb format). Input I read off `/dev/input` with evdev and forwarded into Dart over a platform channel, since key events don't get delivered without a focused window.

Leaving a reference for the framebuffer path in case it saves someone the dead ends I hit: https://pub.dev/packages/flutter_fbdev

**Comment by jwinarske** [comment]
@ardera yes.  I need an ARM engine “SDK” to flush out the BeagleBone black (fbdev).  Everything is there and should work, just needs a little more time with some HW.  Also have an i.MX93 reference board incoming; 64-bit no GPU.

**Comment by jwinarske** [comment]
@ardera Below notes of i.MX93 running with IVI-homescreen.  The software back end also runs on the Beagle Bone Black.

## i.MX93 (no GPU) — wayland-egl vs software backend

**Setup:** wonderous bundle, 1280×720, `IVI_PROFILE` (shared `FrameProfile`,
identical metric both backends), 25 s window, 3 runs each. wayland-egl as a
weston client (llvmpipe); software direct to imx-drm via the dumb sink
(`IVI_SW_SINK=drm-dumb`, weston stopped).

| metric (median of 3) | wayland-egl | software | winner |
| --- | --- | --- | --- |
| fps | 3.98 (3.82–4.04) | 5.36 (5.20–5.45) | software **+35%** |
| CPU (of 2 cores) | ~191% | ~181% | software (weston adds more on top, separate process) |
| RSS | ~290 MB | ~153 MB | software **−47%** |
| cadence | ~2% @60Hz, rest slow(51–100 ms)/idle | 20–30 Hz buckets 27–36% of frames | software (faster per-frame) |

**Conclusion:** on the GPU-less i.MX93, the software backend is clearly better —
~35% higher fps, slightly less CPU (much less system-wide once weston is
counted), and roughly half the memory, with no GL/EGL/compositor dependency.
Both saturate the 2×A55 (~90–95% total), so it's an efficiency win: software
produces more frames per CPU-second by skipping the llvmpipe-GL → EGL → wl_shm →
weston path and rasterizing straight to the dumb buffer.

**Comment by iamsergio** [comment]
There's no winner when going from 4fps to 5fps, it's just 1 frame difference, both unusable for interactive apps.

But maybe worth testing something with less effects than wonderous, then the 35% difference could be more meaningful

**Comment by jwinarske** [comment]
@iamsergio It’s showing worse case; during the test I was clicking the mouse as fast as possible to flip pages.  Any normal use case will only be better.  It works, and all the features are wired up.

**Comment by iamsergio** [comment]
Any idea how many fps while just scrolling a list ?

**Comment by jwinarske** [comment]
> Any idea how many fps while just scrolling a list ?

## i.MX93 IVI-homescreen software backend — list-scroll fps

Validated on **i.MX93** (Yocto Scarthgap - `imx93-11x11-lpddr4x-frdm`) for the
list-scrolling use case using a purpose-built benchmark app.

### Method
- **App:** `scroll_bench` — deterministic `ListView.builder`, 2000 info rows
  (icon + title + two-line detail + trailing value, fixed `itemExtent: 76`).
  Built-in **auto-scroll** flings the list top↔bottom at a constant velocity and
  loops, so the benchmark self-drives and renders identical pixels every run
  (no input injection).
- **Build:** arm64 release AOT via `emb bundle --build --arch arm64 --mode release`.
- **Profiling:** `IVI_SW_PROFILE=1` — the software backend's `FrameProfile`,
  which measures cadence over the **drm-cxx `DrmDumbSink`** present path.
- weston stopped for the run (software backend needs DRM master), restarted after.
- 1280×720 @ 60 Hz, ~45 s sustained scroll.

### Result

| Metric | Value |
| --- | --- |
| Steady-state fps | **~59.0** (per-second samples 58.96–59.03) |
| Session average | **60.03 fps** (2688 frames) |
| Cadence | **90.1%** @ 60 Hz, 9.9% @ 30 Hz |
| Slow / dropped frames | **0** |
| `present_failures` | **0** |

**Vsync-locked at 60 fps with zero slow frames and zero present failures** — the
i.MX93 CPU rasterizes a realistic scrollable info list comfortably over the
drm-cxx software present path.

> Context: wonderous (blur/parallax-heavy, pathological for software raster) runs
> ~5.3 fps software on the same board. A plain info list is the realistic
> list-scroll case and runs at full 60.

<details>
<summary>session summary (verbatim)</summary>

[SoftwareBackend] sink: drm-dumb (device='(probed)', 1280x720@60.00Hz)
[SoftwareBackend] session summary: frames=2688 fps=60.03 mean_interval=16658us max_interval=24256us present_failures=0
[SoftwareBackend] session buckets: 60Hz=2421 (90.1%) 30Hz=266 (9.9%) 20Hz=0 (0.0%) slow=0 (0.0%) idle=0 (0.0%)
</details>

Test Source: https://github.com/toyota-connected/ivi-homescreen/tree/v2.0/test/integration/scroll_bench
emb: https://github.com/toyota-connected/emb_cli

**Comment by jwinarske** [comment]
@iamsergio @ardera
All HW / back end variants are available for PiOS `trixie`/`bookworm` here:
https://github.com/toyota-connected/ivi-homescreen/actions/runs/27874797471

@shwetashah16 IVI-homescreen has first class software support - DRM-mini, fbdev, and file.
